lxd/storage/load.go

```diff
@@ -0,0 +1,122 @@
+package storage
+
+import (
+	"github.com/lxc/lxd/lxd/operations"
+	"github.com/lxc/lxd/lxd/state"
+	"github.com/lxc/lxd/lxd/storage/drivers"
+	"github.com/lxc/lxd/shared/api"
+)
+
+// MockBackend controls whether to run the storage logic in mock mode.
+var MockBackend = false
+
+// CreatePool creates a new storage pool on disk and returns a Pool interface.
+func CreatePool(state *state.State, poolID int64, dbPool *api.StoragePool, op *operations.Operation) (Pool, error) {
+	// Sanity checks.
+	if dbPool == nil {
+		return nil, ErrNilValue
+	}
+
+	// Ensure a config map exists.
+	if dbPool.Config == nil {
+		dbPool.Config = map[string]string{}
+	}
+
+	// Handle mock requests.
+	if MockBackend {
+		pool := mockBackend{}
+		pool.name = dbPool.Name
+		pool.state = state
+		return &pool, nil
+	}
+
+	// Define a function to retrieve a volume ID for a volume Name for use in driver.
+	volIDFunc := func(volType drivers.VolumeType, volName string) (int64, error) {
+		volTypeID, err := VolumeTypeToDBType(volType)
+		if err != nil {
+			return -1, err
+		}
+
+		// TODO add project support in the future by splitting the volName by "_".
+		volID, _, err := state.Cluster.StoragePoolNodeVolumeGetTypeByProject("default", volName, volTypeID, poolID)
+		if err != nil {
+			return -1, err
+		}
+
+		return volID, nil
+	}
+
+	// Load the storage driver.
+	driver, err := drivers.Load(dbPool.Driver, dbPool.Name, dbPool.Config, volIDFunc)
+	if err != nil {
+		return nil, err
+	}
+
+	// Setup the pool struct.
+	pool := lxdBackend{}
+	pool.driver = driver
+	pool.id = poolID
+	pool.name = dbPool.Name
+	pool.state = state
+
+	// Create the pool itself on the storage device..
+	err = pool.create(dbPool, op)
+	if err != nil {
+		return nil, err
+	}
+
+	return &pool, nil
+}
+
+// GetPoolByName retrieves the pool from the database by its name and returns a Pool interface.
+func GetPoolByName(state *state.State, name string) (Pool, error) {
+	// Handle mock requests.
+	if MockBackend {
+		pool := mockBackend{}
+		pool.name = name
+		pool.state = state
+		return &pool, nil
+	}
+
+	// Load the database record.
+	poolID, dbPool, err := state.Cluster.StoragePoolGet(name)
+	if err != nil {
+		return nil, err
+	}
+
+	// Ensure a config map exists.
+	if dbPool.Config == nil {
+		dbPool.Config = map[string]string{}
+	}
+
+	// Define a function to retrieve a volume ID for a volume Name for use in driver.
+	volIDFunc := func(volType drivers.VolumeType, volName string) (int64, error) {
```

Comment:
Given you have two near identical copies of this, maybe make a function which takes in the poolID and return you this function, then use it in both CreatePool and GetPoolByName?

Reply:
Good idea, have implemented.